tests/model/machine_test.py

```diff
@@ -282,6 +283,26 @@ def test_add_meta_overwrite_port(default_device: Machine):
     assert default_device.meta["ports"][(3000, "tcp")] == 5000
     assert result == 4000
 
+def test_add_meta_ulimit(default_device):
+    default_device.add_meta("ulimit", "nofile=1024:2048")
+    assert default_device.meta['ulimits']['nofile'] == {'soft': 1024, 'hard': 2048}
+
+def test_add_meta_ulimit_soft_only(default_device):
+    default_device.add_meta("ulimit", "nofile=1024")
+    assert default_device.meta['ulimits']['nofile'] == {'soft': 1024, 'hard': 1024}
+
+def test_add_meta_ulimit_negative_value(default_device):
+    default_device.add_meta("ulimit", "memlock=-1")
+    assert default_device.meta['ulimits']['memlock'] == {'soft': -1, 'hard': -1}
+
+def test_add_meta_ulimit_invalid_format(default_device):
+    with pytest.raises(MachineOptionError):
+        default_device.add_meta("ulimit", "nofile=1024:2048:4096")
+
+def test_add_meta_ulimit_not_format_exception(default_device):
+    with pytest.raises(MachineOptionError):
+        default_device.add_meta("ulimit", "nofile1024")
+
```

Comment:
Can you add a test with `soft > hard`? (it's the only case not tested)

Reply:
while adding that test i realized i was not covering some edge cases like people giving less than -1 or cases where hard = -1 and limit is a number different from -1 which was not properly handled . I think now i've addressed all cases and also properly added tests.